src/heroic.rs

```diff
@@ -0,0 +1,252 @@
+    fn detect_flatpak_roots(&mut self, _root: &RootsConfig) {
+        // TODO.2022-10-15 heroic: handle games storing stuff in ~/.var/app/com.heroicgameslauncher.hglw
+        // TODO.2022-10-15 heroic: for some games, .var/app/com.heroicgameslauncher.hglw is a root just like $HOME
+    }
+
+    fn detect_legendary_roots(&mut self, root: &RootsConfig) {
+        log::trace!("detect_legendary_roots searching for legendary config...");
+
+        for &legendary_path_candidate in LEGENDARY_PATHS {
+            let legendary_path = StrictPath::new(legendary_path_candidate.to_string());
+            if legendary_path.is_dir() {
+                log::trace!(
+                    "detect_legendary_roots found legendary configuration in {}",
+                    legendary_path.interpret()
+                );
+
+                let mut legendary_installed = legendary_path.as_std_path_buf();
+                legendary_installed.push("installed.json");
+                if legendary_installed.is_file() {
+                    // read list of installed games and call find_game_root for result
+                    if let Ok(ins) = serde_json::from_str::<LegendaryInstalled>(
+                        &std::fs::read_to_string(legendary_installed).unwrap_or_default(),
+                    ) {
+                        ins.0.values().for_each(|game| {
+                            log::trace!("detect_legendary_roots found game {} ({})", game.title, game.game_id);
+                            // process game from GamesConfig
+                            if let Some(sp) = self.find_game_root(
+                                root.path.interpret(),
+                                &game.title,
+                                &game.platform.to_lowercase(),
+                                &game.game_id,
+                            ) {
+                                self.memorize_game_root(&game.title, &sp);
+                            }
+                        });
+                    }
+                } else {
+                    log::trace!(
+                        "detect_legendary_roots no such file '{:?}', legendary probably not used yet... skipping",
+                        legendary_installed
+                    );
+                }
+            }
+        }
+    }
+
+    fn detect_gog_roots(&mut self, root: &RootsConfig) {
+        log::trace!(
+            "detect_gog_roots searching for GOG information in {}",
+            root.path.interpret()
+        );
+
+        // use gog_store/library.json to build map .app_name -> .title
+        let mut game_titles = std::collections::HashMap::<String, String>::new();
+        let gog_library = serde_json::from_str::<GogLibrary>(
+            &std::fs::read_to_string(format!("{}/gog_store/library.json", root.path.interpret())).unwrap_or_default(),
+        );
+        gog_library.unwrap().games.iter().for_each(|game| {
+            game_titles.insert(game.app_name.clone(), game.title.clone());
+        });
```

Comment:
Instead of unwrapping, we can return early if it fails.

Also, we should be able to do something like:

```rust
let game_titles: HashMap<String, String> = gog_library.games.iter().map(|game| {
    (game.app_name.clone(), game.title.clone())
}).collect();
```

Reply:
Replaced all unwrap() with checks for Some